README.md
# Clojure SQL for HBase Phoenix

Clojure SQL for HBase [Phoenix](http://phoenix.apache.org). This
library extends [honeysql](https://github.com/jkk/honeysql) with
additional constructs to support Phoenix-specific queries, such as
upsert, dynamic columns, etc. It facilitates building SQL queries with
clojure data structure.

## Build

[![Build Status](https://travis-ci.org/juvenn/honeysql-phoenix.svg?branch=master)](https://travis-ci.org/juvenn/honeysql-phoenix)
[![Clojars Project](https://img.shields.io/clojars/v/honeysql-phoenix.svg)](https://clojars.org/honeysql-phoenix)

## Getting started

Specify dependencies:

```clj
:dependencies [
  [honeysql-phoenix "0.2.1"]
]
```

In addition to that, a (compatible) phoenix client
(e.g. `phoenix-{version}-client.jar`) should be added to classpath.

## Examples

```clj
(:require [phoenix.db :refer [defdb deftable phoenix] :as db]
          [phoenix.honeysql :refer :all])
```

First of all, define db connection:

```clj
(defdb my-db
  (phoenix {:zk-quorum "127.0.0.1,127.0.0.2:2181"}))
```

Then define table(s) with optional dynamic typed columns:

```clj
(deftable user
  (db/db* my-db)
  (db/table* :user)
  ;; define dynamic columns with its type
  (db/types* :referrer    "VARCHAR(64)"
             :landing_url "VARCHAR(64)"))
```

To insert a row:

```clj
(-> (upsert-into user)
    (values [{:username "jack" :email "jack@example.net"
              :referrer "google.com"}])
    db/exec)
```

In place of `db/exec`, we could invoke `as-sql` to render it as sql
string:

```clj
(-> (upsert-into user)
    (values [{:username "jack" :email "jack@example.net"
              :referrer "google.com"}])
    as-sql)
["UPSERT INTO user (username, email, referrer VARCHAR(64)) VALUES (?, ?, ?)"
 "jack" "jack@example.net" "google.com"]
```

Note that `referrer` is type annotated.

To query rows:

```clj
(-> (select :username :email :referrer :landing_url)
    (from user)
    (where [:= :email "jack@example.net"])
    (limit 1)
    as-sql)
;; manually formatted for ease of reading
["SELECT username, email, referrer, landing_url
  FROM user (referrer VARCHAR(64), landing_url VARCHAR(64))
  WHERE email = ? LIMIT ?"
"jack@example.net" 1]
```

To delete rows:

```clj
(-> (delete-from user)
    (where [:= :email "jack@example.net"])
    as-sql)
["DELETE FROM user WHERE email = ?"
 "jack@example.net"]
```

Atomic update:

```clj
(-> (upsert-into user)
    (values [{:username "jack" :email "jack@example.net"
              :referrer "google.com"}])
    (on-duplicate-key :ignore)
    as-sql)
["UPSERT INTO user (username, email, referrer VARCHAR(64)) VALUES (?, ?, ?) ON DUPLICATE KEY IGNORE"
 "jack" "jack@example.net" "google.com"]

(-> (upsert-into user)
    (values [{:username "jack" :email "jack@example.net"
              :referrer "google.com"}])
    (on-duplicate-key {:referrer "google.com"})
    as-sql)
["UPSERT INTO user (username, email, referrer VARCHAR(64)) VALUES (?, ?, ?)
   ON DUPLICATE KEY UPDATE referrer = ?"
"jack" "jack@example.net" "google.com" "google.com"]
```

For more examples, please refer to [honeysql](https://github.com/jkk/honeysql).

## License

Copyright © 2017, Juvenn Woo.

Distributed under the Eclipse Public License version 1.0.


CHANGES.md

0.2.1
----

* Add `db/avatica` and `db/phoenix` to support different client drivers.

0.2.0
----

* Define table with optional dynamic typed columns
* Syntax for explicitly specifying dynamic columns
* Auto-annotate type in upsert and select columns


test/phoenix/db_test.clj
(ns phoenix.db-test
  (:require [clojure.test :refer :all]
            [phoenix.honeysql :refer :all]
            [phoenix.db :refer [defdb deftable phoenix]
             :as db])
  (:import [phoenix.db Table]))

(defdb default-db
  (phoenix {:zk-quorum "127.0.0.1:2181"}))

(deftest test-deftable
  (deftable my_table)
  (is (= (Table. default-db :my_table)
         my_table))
  (deftable my-table
    (db/table* :my_table))
  (is (= (Table. default-db :my_table)
         my-table))

  (deftable my-table
    (db/table* :my_table)
    (db/types* :field_a "CHAR(64)"
               :field_b "VARCHAR(64)"))
  (is (= {:field_a "CHAR(64)"
          :field_b "VARCHAR(64)"}
         (db/types* my-table))))

(deftable user
  (db/types* :twitter_id "VARCHAR(64)"
             :github_id  "VARCHAR(64)"
             :referrer   "VARCHAR(64)"
             :landing_url "VARCHAR(64)"))

(deftable address
  (db/types* :line2 "VARCHAR(128)"))

(deftest test-exec-raw
  (is (= {:result 43}
         (first (db/exec-raw ["SELECT 42+1 AS result"]))))
  (is (= {:result 43}
         (first (db/exec-raw ["SELECT ? + ? AS result" 42 1]))))
  (is (= {:result -43}
         (first (db/exec-raw ["SELECT ? + ? AS result" 42 1]
                          :row-fn #(update-in % [:result] -)))))
  (testing "DML queries"
    ;; phoenix does not yet support stmt.getGeneratedKeys for upsert statement,
    ;; it always returns number of rows affected.
    (is (= [1]
         (db/exec-raw [(str "UPSERT INTO user (id, username, email, phonenumber) VALUES"
                         " (NEXT VALUE FOR user_id_seq, ?, ?, ?)")
                    "cc5767f0" "cc5767f0@example.net" "25600001234"])))
    (is (= [1]
         (db/exec-raw [(str "DELETE FROM user WHERE email = ?") "cc5767f0@example.net"])))))

(deftest test-exec
  (is (= [1]
         (-> (upsert-into user)
             (columns :id :username :email :referrer)
             (values [[(keyword "NEXT VALUE FOR user_id_seq"), "6e4c580b", "6e4c580b@example.net", "google.com"]])
             (on-duplicate-key :ignore)
             db/exec)))
  (is (= {:username "6e4c580b"
          :email "6e4c580b@example.net"
          :phonenumber nil
          :referrer "google.com"}
         (-> (select :username :email :phonenumber :referrer)
             (from user)
             (where [:= :username "6e4c580b"])
             (limit 1)
             db/exec
             first)))
  (is (= [1]
         (-> (delete-from user)
             (where [:= :username "6e4c580b"])
             db/exec)))
  (is (= nil
         (-> (select :username :email :phonenumber :referrer)
             (from user)
             (where [:= :username "6e4c580b"])
             (limit 1)
             db/exec
             first))))


test/phoenix/avatica_test.clj
(ns phoenix.avatica-test
  (:require [clojure.test :refer :all]
            [phoenix.db :refer [defdb avatica] :as db]))

(defdb avatica-db
  (avatica {:url "http://127.0.0.1:8765"
            :serialization :protobuf}))

(deftest test-avatica-db
  (is (= {:result 43}
         (first (db/exec-raw ["SELECT 42 + 1 AS result"])))))


test/phoenix/honeysql_test.clj
(ns phoenix.honeysql-test
  (:require [clojure.test :refer :all]
            [phoenix.honeysql :refer :all]
            [phoenix.db-test :refer [user address]]))

(deftest test-upsert-into
  (testing "Upsert map form rows should annotate type"
    (is (= [(str "UPSERT INTO user (username, email, referrer VARCHAR(64)) VALUES"
                 " (?, ?, ?),"
                 " (?, ?, ?)")
            "jack" "jack@example.net" "google.com"
            "jill" "jill@example.net" "google.com"]
           (-> (upsert-into user)
               (values [{:username "jack" :email "jack@example.net" :referrer "google.com"}
                        {:username "jill" :email "jill@example.net" :referrer "google.com"}])
               as-sql))))
  (testing "Upsert vector form rows specifying columns"
    (is (= [(str "UPSERT INTO user (username, email, referrer VARCHAR(64))  VALUES"
                 " (?, ?, ?),"
                 " (?, ?, ?)")
            "jack" "jack@example.net" "google.com"
            "jill" "jill@example.net" "google.com"]
           (-> (upsert-into user)
               (columns :username :email :referrer)
               (values [["jack" "jack@example.net" "google.com"]
                        ["jill" "jill@example.net" "google.com"]])
               as-sql))))
  (testing "User specified type is preferred over inferred"
    (is (= [(str "UPSERT INTO user (username, email, referrer CHAR(64))  VALUES"
                 " (?, ?, ?),"
                 " (?, ?, ?)")
            "jack" "jack@example.net" "google.com"
            "jill" "jill@example.net" "google.com"]
           (-> (upsert-into user)
               (columns :username :email [:referrer "CHAR(64)"])
               (values [["jack" "jack@example.net" "google.com"]
                        ["jill" "jill@example.net" "google.com"]])
               as-sql)))))

(deftest test-upsert-on-duplicate-key
  (testing "On duplicate key ignore"
    (is (= [(str "UPSERT INTO user (username, email, phonenumber)  VALUES"
                 " (?, ?, ?)"
                 " ON DUPLICATE KEY IGNORE")
            "jack" "jack@example.net" "15600001234"]
           (-> (upsert-into user)
               (columns :username :email :phonenumber)
               (values [["jack" "jack@example.net" "15600001234"]])
               (on-duplicate-key :ignore)
               as-sql))))
  (testing "On duplicate key update"
    (is (= [(str "UPSERT INTO user (username, email, phonenumber)  VALUES"
                 " (?, ?, ?)"
                 " ON DUPLICATE KEY UPDATE"
                 " username = ?, email = ?, phonenumber = ?")
            "jack"  "jack@example.net"  "15600001234"
            "jack2" "jack2@example.net" "15600001235"]
           (-> (upsert-into user)
               (columns :username :email :phonenumber)
               (values [["jack" "jack@example.net" "15600001234"]])
               (on-duplicate-key {:username "jack2"
                                  :email "jack2@example.net"
                                  :phonenumber "15600001235"})
               as-sql)))))

(deftest test-select
  (testing "No type is annotated"
    (is (= [(str "SELECT username, email FROM user")]
           (-> (select :username :email)
               (from user)
               as-sql))))
  (testing "select fields are auto-annotated"
    (is (= [(str "SELECT username, email, twitter_id AS twitter, github_id AS github"
                 " FROM user (github_id VARCHAR(64), twitter_id VARCHAR(64))")]
           (-> (select :username :email
                       [:twitter_id :twitter]
                       [:github_id  :github])
               (from user)
               as-sql))))
  (testing "Explicitly annotate types"
    (is (= [(str "SELECT username, email, twitter_id AS twitter, github_id AS github"
                 " FROM user (github_id VARCHAR(64), twitter_id CHAR(64))")]
           (-> (select :username :email
                       [:twitter_id :twitter]
                       [:github_id  :github])
               (from [user {:twitter_id "CHAR(64)"}])
               as-sql))))
  (testing "Alias table with explicit annotated types"
    (is (= [(str "SELECT username, email, twitter_id AS twitter, github_id AS github"
                 " FROM user u (github_id VARCHAR(64), twitter_id CHAR(64))")]
           (-> (select :username :email
                       [:twitter_id :twitter]
                       [:github_id  :github])
               (from [user :u {:twitter_id "CHAR(64)"}])
               as-sql))))
  (testing "Select aliased fields are annnotated"
    (is (= [(str "SELECT first_name AS name, addr.line2, addr.zipcode"
                 " FROM user, address addr (line2 CHAR(64))")]
           (-> (select [:first_name :name]
                       :addr.line2
                       :addr.zipcode)
               (from user [address :addr {:line2 "CHAR(64)"}])
               as-sql))))
  (testing "Subquery are annotated"
    (is (= [(str "SELECT referrer, user_cnt FROM ("
                 "SELECT referrer, count(user_id) AS user_cnt FROM"
                 " user (referrer VARCHAR(64))"
                 " GROUP BY referrer"
                 ") WHERE user_cnt > ?") 5]
           (-> (select :referrer :user_cnt)
               (from (-> (select :referrer [:%count.user_id :user_cnt])
                         (from user)
                         (group :referrer)))
               (where [:> :user_cnt 5])
               as-sql)))))



src/phoenix/db.clj
(ns phoenix.db
  (:require [honeysql.format :as fmt]
            [honeysql.core :as sql]
            [clojure.string :as str]
            [clojure.java.jdbc :as jdbc]))

(def ^:dynamic *default-db* (atom nil))

(defn phoenix
  "Create an phoenix jdbc driver spec.

  :zk-quorum  Zookeeper quorum url, including port, default `127.0.0.1:2181`
  :zk-parent  Zookeeper parent, default `/hbase`
  :principal  Kerberos principal, optional
  :keytab     Kerberos keytab, optional

  See Phoenix [FAQ](http://phoenix.apache.org/faq.html#What_is_the_Phoenix_JDBC_URL_syntax)."
  [spec]
  (-> spec
      (update :dbtype #(or % :phoenix))
      (update :connection-uri
              (fn [uri]
                (or uri
                    (->> (map spec [:zk-parent :principal :keytab])
                         (remove nil?)
                         (str/join ":")
                         (str "jdbc:phoenix:"
                              (or (:zk-quorum spec) "127.0.0.1:2181")
                              ":")))))))

(defn avatica
  "Create an avatica jdbc driver spec.

  :url            Full url of avatica server (or phoenix query server), e.g.:
                  http://127.0.0.1:8765
  :serialization  Serialization protocol to communicate with server,
                  `json` or `protobuf`.

  For more options, please see:
  https://calcite.apache.org/avatica/docs/client_reference.html"
  [spec]
  (-> spec
      (update :dbtype #(or % :avatica))
      (update :connection-uri
              (fn [uri]
                (or uri
                    (reduce-kv (fn [uri k v]
                                 (if v
                                   (str uri (name k) "=" (name v) ";")
                                   uri))
                               "jdbc:avatica:remote:"
                               spec))))))

(defmacro defdb
  "
  Define a db spec and set it as default.
  "
  [db-name spec]
  `(let [db# ~spec]
     (defonce ~db-name db#)
     (reset! *default-db* ~db-name)))

(defrecord Table [db table]
  fmt/ToSql
  (to-sql [_]
    (fmt/to-sql (keyword table))))

(defmacro deftable
  "Define a table with dynamic columns."
  [table & forms]
  `(let [spec# {:db @*default-db*
                :table (keyword '~table)}]
     (def ~table
       (-> (map->Table spec#)
           ~@forms))))

(defn table?
  "Test if it is a table instance."
  [table]
  (instance? Table table))

(defn db*
  "Return db spec, or set db sepc of a table."
  ([table]
   (when (table? table)
     (:db table)))
  ([^Table table db]
   (assoc table :db db)))

(defn table*
  "Return table name, or set table name."
  ([table]
   (if (table? table)
     (:table table)
     table))
  ([^Table table name]
   (assoc table :table name)))

(defn types*
  "Return column type map, or set column types of a table."
  ([table]
   (when (table? table)
     (:types table)))
  ([^Table table types]
   (update table :types merge types))
  ([^Table table k v & types]
   (-> table
       (assoc-in [:types k] v)
       (update :types into (map vec (partition 2 types))))))

(defn- select-query? [^String query]
  (or (str/starts-with? query "SELECT")
      (str/starts-with? query "select")))

(defn exec-raw
  "Execute raw sql, either in query (select) or update mode. In the query mode,
   options such as :row-fn, :result-fn could be supplied to transform queried
   result, return seq of rows. E.g.:

  (exec-raw [\"SELECT * FROM user WHERE username = ? LIMIT ?\" \"jack\" 1]
            :db my-db
            :row-fn identity)

  While in update mode, return seq of number of rows affected. Phoenix does
  not yet support generated keys in update mode.

  See also jdbc/query, jdbc/execute!.
  "
  [query & {:keys [db] :as opts}]
  (if (select-query? (first query))
    (jdbc/query    (or db @*default-db*) query opts)
    (jdbc/execute! (or db @*default-db*) query opts)))

(defn exec
  "Render sqlmap as sql string, then execute it.
  See also: exec-raw."
  [sqlmap & {:as opts}]
  (let [table (or (:upsert-into sqlmap)
                  (:delete-from sqlmap))]
    (exec-raw (sql/format sqlmap)
              :db (db* (or table
                           (#(if (sequential? %) (first %) %)
                            (first (:from sqlmap))) )))))


src/honeysql/phoenix/core.clj
(ns honeysql.phoenix.core
  (:require [clojure.string :as str]
            [honeysql.format :as fmt]
            [honeysql.core :as sql]
            [honeysql.helpers :refer [defhelper]
             :as h]))

(def ^:dynamic *default-db* nil)

(defn add-connection-uri
  "Build and add connection-uri to spec if not present."
  [{:keys [quorum zk-path] :as spec}]
  (if (:connection-uri spec)
    spec
    (assoc spec :connection-uri
           (str "jdbc:phoenix:"
                (or quorum "127.0.0.1:2181")
                ":"
                (or zk-path "/hbase")))))

(defmacro defdb
  "
  Define a connection spec in the form of jdbc spec. The first def
  will be used by default for queries where no db were specified. The
  following options can be specified in the spec:

  :connection-uri  If present will be used to connect to jdbc regardless
                   of :quorum and :zk-path.
  :quorum          zookeeper quorum, it should contain port. Default to
                   be `127.0.0.1:2181`.
  :zk-path         hbase's zookeeper znode parent, default to be
                   `/hbase`.

  See java.jdbc.
  "
  [db-name spec]
  `(let [spec# (add-connection-uri ~spec)]
     (defonce ~db-name spec#)))

(defrecord Table [db table columns dynamic]
  fmt/ToSql
  (to-sql [_]
    (fmt/to-sql (keyword table))))

(defmacro deftable
  "Defind a table with dynamic columns."
  [table spec]
  `(let [spec# (-> ~spec
                   (update-in [:db] #(or % *default-db*))
                   (update-in [:table] #(or % (keyword '~table))))]
     (def ~table (map->Table spec#))))

(defn table-name
  [table]
  (if (instance? Table table)
    (:table table)
    table))

(defn table-db [table]
  (if (instance? Table table)
    (:db table *default-db*)
    *default-db*))

;; ----

(defmethod fmt/format-clause :on-duplicate-key [[op values] sqlmap]
  (if (= :ignore values)
    "ON DUPLICATE KEY IGNORE"
    (str "ON DUPLICATE KEY UPDATE "
         (fmt/comma-join (for [[k v] values]
                           (str (fmt/to-sql k) " = " (fmt/to-sql v)))))))

(fmt/register-clause! :on-duplicate-key 225) ;; after :values

;; Reset :columns to be no-op, which should be handled at
;; :from, :upsert-into, :insert-into.
(defmethod fmt/format-clause :columns [[_ fields] sqlmap]
  "")

(defn- annotate-type
  "Annotate type to column if it's present."
  [type col]
  (if type
    [col type]
    col))

(defn- attach-types
  "Attach type definitions to columns."
  [type-defs columns]
  (map
   #(annotate-type (get type-defs %) %)
   columns))

(defn- refn-alias
  "Split a reference into ref and its alias (nil if not present)."
  [table]
  (if (sequential? table)
    [(first table) (second table)]
    [table nil]))

(defn format-table
  "Format table with optional typed columns: TEST_TABLE(a, x INTEGER ...)"
  [table columns]
  (let [[table aliaz] (refn-alias table)
        cols (if (instance? Table table)
               (attach-types (:dynamic table) columns)
               columns)]
    (str (fmt/to-sql table)
         (when-not (empty? cols)
           (str " (" (fmt/comma-join (map fmt/to-sql cols)) ")"))
         (when aliaz
           (str " " (fmt/to-sql aliaz))))))

(defmethod fmt/format-clause :upsert-into [[op table] sqlmap]
  (str "UPSERT INTO " (format-table table (:columns sqlmap))))

;; Phoenix does not support insert-into, define it anyway.
(defmethod fmt/format-clause :insert-into [[op table] sqlmap]
  (str "INSERT INTO " (format-table table (:columns sqlmap))))

(fmt/register-clause! :upsert-into 45)       ;; before :select

(defn split-qualifier
  "Split qualified columns into qualifier and name, return tuple.
  Non-qualified will have qualifier `*`."
  [s]
  (let [xs (str/split s #"\.")]
    (if (= 1 (count xs))
      ["*" (first xs)]
      xs)))

(defn group-qualified-names
  "Group qualified names into qualifier and its names. E.g.:

    => (group-qualified-names [:a \"test.b\" :test.c])
    {:* (:a) :test (:c :b)}

  Note non-qualified names are grouped under :*.
  "
  [names]
  (->> names
       (filter #(or (keyword? %) (string? %)))
       (map (comp split-qualifier name))
       (reduce (fn [m [k v]]
                 (update m (keyword k) conj (keyword v)))
               {})))

(defmethod fmt/format-clause :from [[_ tables] sqlmap]
  (let [cols (:columns sqlmap) ;; user specified columns
        table-cols (->> cols
                        (into (:select sqlmap))
                        (map (comp first refn-alias))
                        group-qualified-names)]
    (str "FROM "
         (fmt/comma-join
          (for [table tables]
            (let [[t alias] (refn-alias table)]
              (condp instance? t
                Table
                (format-table table
                              (->> (get table-cols :*)
                                   ;; full qualified columns
                                   (concat (get table-cols (table-name t)))
                                   ;; alias qualified columns
                                   (concat (get table-cols (keyword alias)))
                                   set
                                   ;; only handle dynamic columns
                                   (filter #(contains? (:dynamic t) %))))
                String
                (format-table table cols)
                clojure.lang.Keyword
                (format-table table cols)
                ;; subquery etc.
                (fmt/to-sql table))))))))

;; TODO: support join

;; ---- Define helpers

(defhelper on-duplicate-key [m args]
  (assoc m :on-duplicate-key (if (sequential? args)
                               (first args)
                               args)))

(defhelper upsert-into [m table]
  (assoc m :upsert-into (if (sequential? table) (first table) table)))

(defmacro upsert-into!
  "Exectue and upsert records. E.g.:

  (upsert! web-stat
         (columns :a :b [:dynamic_col :int] ...)
         (values ...)
         (on-duplicate-key :ignore))
  "
  ([table & forms]
   (let [db# (table-db ~table)]
     `(-> (upsert-into (table-name ~table))
          ~@forms
          (update :columns
                  #(attach-types (:dynamic ~table) %))))))

(defn from-clause
  "Return clause if it's form of `(from ...)`, false otherwise."
  [clause]
  (and (list? clause)
       (= #'h/from (resolve (first clause)))
       clause))

(defmacro select!
  "Select, execute, and return result. E.g.:

  (select! :a :b
         (from web-stat)
         (where ...))
  "
  ([col & forms]
   ;; TODO: support alias and multiple tables
   (let [[cols# clauses#] (split-with (comp not list?) forms)
         from-clause# (some from-clause clauses#)
         clauses# (doall (remove from-clause clauses#))
         table# (second from-clause#)
         db# (table-db table#)
         cols# (cons col cols#)
         dyna-cols# (->> cols#
                         (select-keys (:dynamic (var-get (resolve table#)) {}))
                         vec)]
     `(-> (h/select ~@cols#)
          (h/from (table-name ~table#))
          ~@clauses#
          (update :columns
                  #(if (empty? %)
                     ~dyna-cols#
                     %))))))


(defmacro delete! [])

; :columns (:a :b :c [:d :int] :e)
; :columns ([:a :b :c [:d :int] :e])
; (-> (upsert-into web-stat)
;     (columns ...)
;     (values ...)
;     upsert!)

; for upsert!:
; 1. columns not present, do nothing
; 2. columns present, parse and add dynamic type

; (-> (select [...])
;     (from web-stat)
;     (where ...)
;     select!)
;
; for select!:
; 1. :* present, add all dynamic columns
; 2. columns present, parse and add dynamic type


src/phoenix/honeysql.clj
(ns phoenix.honeysql
  (:require [clojure.string :as str]
            [potemkin :refer [import-vars]]
            [phoenix.db :as db]
            [honeysql.format :as fmt]
            [honeysql.core :as sql]
            [honeysql.helpers :refer [defhelper]
             :as h]))

(import-vars
 [honeysql.helpers
  columns
  values
  modifiers
  select
  from
  where
  join
  left-join
  right-join
  full-join
  group
  having
  order-by
  limit
  offset
  delete-from])

(def as-sql
  "Format sqlmap as sql string, return vector of sql string and params.
  See also honeysql.core/format."
  #'sql/format)

(fmt/register-clause! :on-duplicate-key 225) ;; after :values

(defhelper on-duplicate-key [m args]
  (assoc m :on-duplicate-key (if (sequential? args)
                               (first args)
                               args)))

(defmethod fmt/format-clause :on-duplicate-key [[op values] sqlmap]
  (if (= :ignore values)
    "ON DUPLICATE KEY IGNORE"
    (str "ON DUPLICATE KEY UPDATE "
         (fmt/comma-join (for [[k v] values]
                           (str (fmt/to-sql k) " = " (fmt/to-sql v)))))))

;; Reset :columns to be no-op
(defmethod fmt/format-clause :columns [[_ fields] sqlmap]
  "")

(defn- ref-alias
  "Partition a form into ref and alias, assuming [ref alias] form."
  [form]
  (if (sequential? form)
    [(first form) (second form)]
    [form nil]))

(defn- format-columns [cols]
  (fmt/paren-wrap (fmt/comma-join (map fmt/to-sql cols))))

(defmethod fmt/format-clause :values [[_ values] _]
  (if (sequential? (first values))
    ;; value in vector form
    (str "VALUES "
         (fmt/comma-join
          (for [xs values] (format-columns xs))))
    ;; value in map form
    (let [ks (keys (first values))]
      (str "VALUES "
           (fmt/comma-join
            (for [m values]
              (format-columns (map (partial get m) ks))))))))

(defn- format-table-cols
  "Format table with (optionally-typed) columns."
  [table cols]
  (let [[table alias] (if (sequential? table)
                        table
                        [table])]
    (str (fmt/to-sql table)
         (when alias
           (str " " (fmt/to-sql alias)))
         (when-let [cols (seq cols)]
           (str " " (format-columns cols))))))

(defn- annotate-types
  "Annotate cols with types, return seq of cols with optional types.
   E.g.:

  ```
  => (annotate-types user [:username :email :referrer])
  (:username :email [:referrer \"VARCHAR(64)\"])
  ```
  "
  [table cols]
  (if-not (db/table? table)
    cols
    (let [get-type (partial get (db/types* table))]
      (map #(if-let [type* (get-type %)]
              [% type*]
              %)
           (seq cols)))))

(fmt/register-clause! :upsert-into 45)       ;; before :select

(defmethod fmt/format-clause :upsert-into [[op table] sqlmap]
  (let [cols (or (:columns sqlmap)
                 (when-let [row (first (:values sqlmap))]
                   (when (map? row)
                     (keys row))))]
    (->> (annotate-types table cols)
         (format-table-cols table)
         (str "UPSERT INTO "))))

(defhelper upsert-into [m table]
  (assoc m :upsert-into (if (sequential? table) (first table) table)))

(defn- split-qual-col
  "Split column name into tuple of [qual name], keywordized. Unqualified columns are
   grouped under :_."
  [term]
  (let [[qual col] (str/split (name term) #"\.")]
    (if col
      [(keyword qual) (keyword col)]
      [:_ (keyword qual)])))

(defn- infer-qual-col
  "Infer qual(ifier) and column from a column related expr, return
   qual col pair."
  [expr]
  (cond
    (string? expr) (split-qual-col expr)
    (keyword? expr) (split-qual-col expr)
    (and (vector? expr)
         (= 2 (count expr)))
    (split-qual-col (first expr))
    :else nil))

(defn- group-qual-col
  "Group qual-col pair by qualifier, keywordized."
  [pairs]
  (reduce
   (fn [m [k v]]
     (if v
       (update m (keyword k) (fnil conj []) v)
       m))
   {}
   pairs))

(defmethod fmt/format-clause :from [[_ tables] sqlmap]
  (let [select-cols (->> (:select sqlmap)
                         (map infer-qual-col)
                         (group-qual-col))]
    (str "FROM "
         (fmt/comma-join
          (for [table-ref tables]
            (let [[table alias types] (if (sequential? table-ref)
                                         table-ref
                                         [table-ref])
                  [alias types] (if (map? alias)
                                  [nil alias]
                                  [(keyword alias) types])]
              (cond
                ;; no type to infer about
                (keyword? table)
                (format-table-cols [table alias] types)

                (db/table? table)
                (->> select-cols
                     ((juxt :_ (or alias :_null) (db/table* table)))
                     (reduce into #{})
                     (select-keys (db/types* table))
                     ;; user specified type is preferred over inferred
                     (#(merge %2 %1) types)
                     (format-table-cols [table alias]))

                ;; subquery etc
                :else
                (fmt/to-sql table-ref))))))))

